db/migrate/20220722101202_add_reference_to_planning_applications.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+class AddReferenceToPlanningApplications < ActiveRecord::Migration[6.1]
+  class PlanningApplication < ApplicationRecord
+    def set_reference
+      self.reference = [
+        created_at_year,
+        application_number,
+        application_type_code
+      ].join("-")
+    end
+
+    private
+
+    enum application_type: { lawfulness_certificate: 0, full: 1 }
+
+    def application_number
+      self[:application_number].to_s.rjust(5, "0")
+    end
+
+    def created_at_year
+      created_at.strftime("%y")
+    end
+
+    def application_type_code
+      I18n.t("application_type_codes.#{application_type}.#{work_status}")
+    end
+  end
+
+  def up
+    add_column :planning_applications, :reference, :string
+    add_index :planning_applications, "lower(reference)"
+
+    add_index(
+      :planning_applications,
+      %i[reference local_authority_id],
+      unique: true
+    )
+
+    PlanningApplication.all.find_each do |planning_application|
+      planning_application.set_reference
+      planning_application.save!
+    end
+  end
```

Comment:
do we need to add a not null constraint to reference too?

Reply:
But then we'd need to update `after_create :set_reference` to `before_create :set_reference`

Reply:
So I tried this and then realised the (I think) fatal flaw...we use `created_at` in the `reference`!

Reply:
ahhh haha ok

Reply:
I guess we'd use `Time.current` ?

Reply:
Actually yes, I don't see why not! It's only the year, after all, and the changes of someone creating a planning application at the spot of midnight on new years eve and the created at falling those crucial microseconds later is...probably non existent right? 🤣

Reply:
haha what a scenario 😂

Reply:
in that case (in narnia) we could just do a query to see if created at year is different to any planning application ref year and update manually as there is no unique index on the year itself 😄

Reply:
yes, I'm sure if the day ever comes we can handle it 😁